aten/src/ATen/native/cpu/DistanceOpsKernel.cpp

```diff
@@ -0,0 +1,192 @@
+#include "ATen/native/cpu/ReduceOpsKernel.h"
+
+#include <numeric>
+#include <iterator>
+#include <algorithm>
+
+#include "ATen/Dispatch.h"
+#include "ATen/Parallel.h"
+#include "ATen/core/optional.h"
+#include "ATen/cpu/vec256/vec256.h"
+
+namespace at { namespace native { namespace {
+
+template<typename scalar_t>
+struct PDist {
+
+  static inline scalar_t sign(scalar_t val) {
+    return (0 < val) - (val < 0);
+  }
+
+  static scalar_t zdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += *a != *b;
+    }
+    return result;
+  }
+
+  static scalar_t odist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += std::abs(*a - *b);
+    }
+    return result;
+  }
+
+  static scalar_t tdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      scalar_t diff = *a - *b;
+      result += diff * diff;
+    }
+    return std::sqrt(result);
+  }
+
+  static scalar_t pdist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result += std::pow(std::abs(*a - *b), p);
+    }
+    return std::pow(result, 1.0 / p);
+  }
+
+  static scalar_t idist_calc(const scalar_t * a, const scalar_t * b, const int64_t size, const int64_t stride, const double p) {
+    scalar_t result = 0.0;
+    for (int64_t i = 0; i != size; i += 1, a += stride, b += stride) {
+      result = std::max(result, std::abs(*a - *b));
+    }
+    return result;
+  }
+
+  template <scalar_t (*F)(const scalar_t *, const scalar_t *, const int64_t, const int64_t, const double)>
+  static void run_parallel(Tensor& result, const Tensor& self, const double p) {
+    auto res_ = result.data<scalar_t>();
+    auto self_ = self.data<scalar_t>();
+    int64_t n = self.size(0);
+    int64_t m = self.size(1);
+    int64_t ns = self.stride(0);
+    int64_t ms = self.stride(1);
+
+    int64_t combs = n * (n - 1) / 2;
+    parallel_for(0, combs, 1, [=](int64_t k, int64_t end) {
+      float n2 = n - .5;
+      // The -1 accounts for floating point truncation issues
+      int64_t i = (int64_t) ((n2 - std::sqrt(n2 * n2 - 2 * k - 1)));
```

Comment:
Nit: prefer `static_cast<int64_t>(...)`

Reply:
Also, bluuuugh, do we really have to do a `sqrt` to do index computation...

Reply:
The only other solutions I can think of are to ahead of time:
```
vector<int64_t> lookup();
lookup.reserve(combs);
for (int64_t i = 0; i != n - 1; ++i) {
  for (int64_t j = i + 1; j != n; ++j) {
    lookup.push(i);
  }
}
```

and then later `int64_t i = lookup[k]`.

alternatively, we can ignore the startup time and memory for an O(n) computation:
```
int64_t kc = k;
int64_t i = 0;
for (int64_t kc = k, step = n - 1; kc >= step; i += 1, kc -= step, step -= 1) {}
```

This might be faster than the sqrt, but it's unclear. If you have a preference I can change it, but I doubt this is ever a bottleneck.

Reply:
It's fine.